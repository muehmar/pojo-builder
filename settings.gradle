rootProject.name = "pojo-builder"

include 'core'
include 'core-annotations'
include 'example'
include 'example-java17'

dependencyResolutionManagement {
    versionCatalogs {
        libs {
            version('pojobuilder', '1.6.0')
            library('pojobuilder-core', 'io.github.muehmar','pojo-builder').versionRef('pojobuilder')
            library('pojobuilder-annotations', 'io.github.muehmar','pojo-builder-annotations').versionRef('pojobuilder')

            library('codeGenerator', 'io.github.muehmar:code-generator:0.22.0')

            version('junit', '5.10.0')
            library('junit-api', 'org.junit.jupiter', 'junit-jupiter-api').versionRef('junit')
            library('junit-engine', 'org.junit.jupiter', 'junit-jupiter-engine').versionRef('junit')
            library('junit-params', 'org.junit.jupiter', 'junit-jupiter-params').versionRef('junit')

            version('assertj', '3.26.3')
            library('assertj', 'org.assertj', 'assertj-core').versionRef('assertj')

            bundle('junit', ['junit-api', 'junit-engine', 'junit-params', 'assertj'])

            library('jakarta-annotationApi', 'jakarta.annotation:jakarta.annotation-api:2.1.1')

            library('google-autoservice', 'com.google.auto.service:auto-service:1.1.1')
            library('joor', 'org.jooq:joor-java-8:0.9.15')

            library('lombok', 'org.projectlombok:lombok:1.18.32')

            library('snapshotTesting','io.github.origin-energy:java-snapshot-testing-junit5:4.0.8')

            library('slf4j', 'org.slf4j:slf4j-api:2.0.7')

            plugin('release', 'net.researchgate.release').version('2.8.1')
            plugin('versions', 'com.github.ben-manes.versions').version('0.51.0')
        }
    }
}