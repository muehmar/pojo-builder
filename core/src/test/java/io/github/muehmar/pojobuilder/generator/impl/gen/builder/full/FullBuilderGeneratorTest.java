package io.github.muehmar.pojobuilder.generator.impl.gen.builder.full;

import static io.github.muehmar.codegenerator.writer.Writer.javaWriter;
import static org.assertj.core.api.Assertions.assertThat;

import au.com.origin.snapshots.Expect;
import au.com.origin.snapshots.annotations.SnapshotName;
import io.github.muehmar.codegenerator.Generator;
import io.github.muehmar.codegenerator.writer.Writer;
import io.github.muehmar.pojobuilder.annotations.FullBuilderFieldOrder;
import io.github.muehmar.pojobuilder.generator.PojoFields;
import io.github.muehmar.pojobuilder.generator.Pojos;
import io.github.muehmar.pojobuilder.generator.model.Pojo;
import io.github.muehmar.pojobuilder.generator.model.settings.PojoSettings;
import io.github.muehmar.pojobuilder.snapshottesting.SnapshotTest;
import org.junit.jupiter.api.Test;

@SnapshotTest
class FullBuilderGeneratorTest {
  private Expect expect;

  @Test
  @SnapshotName("declarationOrder")
  void fullBuilder_when_declarationOrder_then_correctOutput() {
    final Generator<Pojo, PojoSettings> generator = FullBuilderGenerator.fullBuilderGenerator();

    final Pojo pojo =
        Pojos.fromFields(
            PojoFields.requiredId(), PojoFields.optionalName(), PojoFields.requiredMap());
    final Writer writer =
        generator.generate(
            pojo,
            PojoSettings.defaultSettings()
                .withFullBuilderFieldOrder(FullBuilderFieldOrder.DECLARATION_ORDER),
            javaWriter());

    expect.toMatchSnapshot(writer.asString());
  }

  @Test
  @SnapshotName("requiredFirstOrder")
  void fullBuilder_when_requiredFirstOrder_then_correctOutput() {
    final Generator<Pojo, PojoSettings> generator = FullBuilderGenerator.fullBuilderGenerator();

    final Pojo pojo =
        Pojos.fromFields(
            PojoFields.requiredId(), PojoFields.optionalName(), PojoFields.requiredMap());
    final Writer writer =
        generator.generate(
            pojo,
            PojoSettings.defaultSettings()
                .withFullBuilderFieldOrder(FullBuilderFieldOrder.REQUIRED_FIELDS_FIRST),
            javaWriter());

    expect.toMatchSnapshot(writer.asString());
  }

  @Test
  @SnapshotName("genericPojo")
  void fullBuilder_when_genericPojo_then_correctOutput() {
    final Generator<Pojo, PojoSettings> generator = FullBuilderGenerator.fullBuilderGenerator();

    final Writer writer =
        generator.generate(Pojos.genericSample(), PojoSettings.defaultSettings(), javaWriter());

    expect.toMatchSnapshot(writer.asString());
  }

  @Test
  void fullBuilderGenerator_when_disabledFullBuilder_then_noOutput() {
    final Generator<Pojo, PojoSettings> generator = FullBuilderGenerator.fullBuilderGenerator();
    final Writer writer =
        generator.generate(
            Pojos.sample(),
            PojoSettings.defaultSettings().withFullBuilderEnabled(false),
            javaWriter());

    assertThat(writer.asString()).isEqualTo("");
  }
}
