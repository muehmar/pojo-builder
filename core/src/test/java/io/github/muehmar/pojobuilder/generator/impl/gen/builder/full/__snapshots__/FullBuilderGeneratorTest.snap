declarationOrder=[
public static final class FullBuilder0 {
  private final Builder builder;

  private FullBuilder0(Builder builder) {
    this.builder = builder;
  }

  public FullBuilder1 id(Integer id) {
    return new FullBuilder1(builder.id(id));
  }
}

public static final class FullBuilder1 {
  private final Builder builder;

  private FullBuilder1(Builder builder) {
    this.builder = builder;
  }

  public FullBuilder2 name(String name) {
    return new FullBuilder2(builder.name(name));
  }

  public FullBuilder2 name(Optional<String> name) {
    return new FullBuilder2(builder.name(name));
  }
}

public static final class FullBuilder2 {
  private final Builder builder;

  private FullBuilder2(Builder builder) {
    this.builder = builder;
  }

  public FullBuilder3 someMap(Map<String,List<String>> someMap) {
    return new FullBuilder3(builder.someMap(someMap));
  }
}

public static final class FullBuilder3 {
  private final Builder builder;

  private FullBuilder3(Builder builder) {
    this.builder = builder;
  }

  public Customer build() throws IOException {
    return builder.build();
  }

  /**
   * Creates a new instance of {@link CustomerBuilder} where it is guaranteed that
   * all required fields are already set.
   */
  public CustomerBuilder toBuilder() {
    return builder.toBuilder();
  }
}
]


genericPojo=[
public static final class FullBuilder0<T extends List<String>, S> {
  private final Builder<T, S> builder;

  private FullBuilder0(Builder<T, S> builder) {
    this.builder = builder;
  }

  public FullBuilder1<T, S> id(String id) {
    return new FullBuilder1<>(builder.id(id));
  }
}

public static final class FullBuilder1<T extends List<String>, S> {
  private final Builder<T, S> builder;

  private FullBuilder1(Builder<T, S> builder) {
    this.builder = builder;
  }

  public FullBuilder2<T, S> data(T data) {
    return new FullBuilder2<>(builder.data(data));
  }
}

public static final class FullBuilder2<T extends List<String>, S> {
  private final Builder<T, S> builder;

  private FullBuilder2(Builder<T, S> builder) {
    this.builder = builder;
  }

  public FullBuilder3<T, S> additionalData(S additionalData) {
    return new FullBuilder3<>(builder.additionalData(additionalData));
  }

  public FullBuilder3<T, S> additionalData(Optional<S> additionalData) {
    return new FullBuilder3<>(builder.additionalData(additionalData));
  }
}

public static final class FullBuilder3<T extends List<String>, S> {
  private final Builder<T, S> builder;

  private FullBuilder3(Builder<T, S> builder) {
    this.builder = builder;
  }

  public Customer<T, S> build() throws IOException {
    return builder.build();
  }

  /**
   * Creates a new instance of {@link CustomerBuilder} where it is guaranteed that
   * all required fields are already set.
   */
  public CustomerBuilder<T, S> toBuilder() {
    return builder.toBuilder();
  }
}
]


requiredFirstOrder=[
public static final class FullBuilder0 {
  private final Builder builder;

  private FullBuilder0(Builder builder) {
    this.builder = builder;
  }

  public FullBuilder1 id(Integer id) {
    return new FullBuilder1(builder.id(id));
  }
}

public static final class FullBuilder1 {
  private final Builder builder;

  private FullBuilder1(Builder builder) {
    this.builder = builder;
  }

  public FullBuilder2 someMap(Map<String,List<String>> someMap) {
    return new FullBuilder2(builder.someMap(someMap));
  }
}

public static final class FullBuilder2 {
  private final Builder builder;

  private FullBuilder2(Builder builder) {
    this.builder = builder;
  }

  public FullBuilder3 name(String name) {
    return new FullBuilder3(builder.name(name));
  }

  public FullBuilder3 name(Optional<String> name) {
    return new FullBuilder3(builder.name(name));
  }
}

public static final class FullBuilder3 {
  private final Builder builder;

  private FullBuilder3(Builder builder) {
    this.builder = builder;
  }

  public Customer build() throws IOException {
    return builder.build();
  }

  /**
   * Creates a new instance of {@link CustomerBuilder} where it is guaranteed that
   * all required fields are already set.
   */
  public CustomerBuilder toBuilder() {
    return builder.toBuilder();
  }
}
]